File: ios/KaleidoscopeModule/gpu/Ingest.swift
// Ingest: the ONE place camera orientation is normalized on iOS.
//
// ===== Why this exists (read before "fixing" any effect's orientation) =====
//
// The camera hands us an RTCVideoFrame whose CVPixelBuffer is raw sensor-space
// (typically a LANDSCAPE NV12 buffer) plus a separate `frame.rotation` (0/90/
// 180/270) that the display would otherwise apply downstream. The old iOS code
// ingested the buffer with NO rotation, so the "original" BGRA texture was raw
// landscape; every effect then re-corrected orientation itself (transform ops
// conjugated screen<->buffer, the background pre-oriented through an extra
// transform pass, FrameBridge preserving frame.rotation so the consumer rotated
// the whole composite). That per-effect cascade is what this module removes.
//
// `ingest(...)` below folds `frame.rotation` INTO the CoreImage NV12->BGRA
// render, so the single ingest produces a DISPLAY-UPRIGHT "original" BGRA
// texture. The original buffer is sized with display dims (buffer dims swapped
// on a 90/270 frame), so the upright image fills it without clamping. Effects
// then emit `rotation 0` (the pixels are already upright; nothing downstream
// should re-rotate them). This mirrors android/.../gpu/Ingest.kt, which folds
// the display rotation into the OES->2D pass so its FBO is display-upright.
//
// After ingest the "original" texture is canonical: every downstream pass
// (transform mat2, background cover-fit + composite, blur, the segmentation mask
// downscale) reads an already-upright frame and applies NO further orientation
// correction.
//
// ===== Two SEPARATE concerns; do not conflate them =====
//
//   (1) DISPLAY ROTATION (this file). The camera-vs-display rotation. Lives
//       here, once. `ROTATION_DIRECTION` is the single sign knob.
//
//   (2) METAL PER-PASS V-FLIP (NOT this file). Every Metal render-to-texture
//       pass in this module flips vertically in buffer space, because the
//       transpiled spirv-cross passthrough vertex does not negate
//       gl_Position.y (see MetalRenderer header). That is a RENDER-PASS-COUNT
//       parity property, independent of the camera. The composite samples the
//       original directly in its single pass, so a single-pass effect's
//       foreground lands correct with no V term; blur's background travels
//       through extra ping-pong passes (odd parity) and so still needs its own
//       V term in the blur layer pass. The ingest is a CoreImage render, not a Metal
//       pass, so it does NOT participate in (2): it must not add a V-flip to
//       "fix" a downstream pass-parity issue. If a screenshot is upside-down,
//       decide which concern it is before touching anything.
//
// ===== Calibration (device verifies; may need one literal edit) =====
//
// ROTATION_DIRECTION below is the single sign that decides whether the display
// rotation is applied as +R or -R about the image center. If the person /
// background comes out rotated the wrong way (e.g. 90 off, or upside down on a
// 180 device), flip this sign. Nothing else in the orientation story should
// ever need touching. INGEST_V_FLIP is provided ONLY as an escape hatch in case
// the CoreImage render orientation differs from expectation on device; it should
// stay `false` (the per-pass V parity is handled downstream, not here).

import Foundation
import CoreVideo
import CoreGraphics

enum Ingest {
  // The ONE orientation calibration knob. +1 applies the display rotation as
  // +frame.rotation about the image center (CoreImage's CCW-positive
  // convention). Flip to -1 if a portrait-device screenshot shows the whole
  // frame rotated the wrong way. This is the only place orientation direction
  // is decided. (Android's analogue is Ingest.ROTATION_DIRECTION.)
  static let ROTATION_DIRECTION: CGFloat = 1.0

  // De-mirror knob. The front camera delivers a HORIZONTALLY MIRRORED buffer
  // (its CVPixelBuffer is selfie-mirrored), so a pure-rotation ingest produces a
  // mirrored "original": Flip X cancels to a no-op, Flip Y reads as a 180, and
  // the debug-grid text reads BACKWARDS. Android (MLKit, no mirror) and the web
  // reference both have a NON-mirrored canonical frame. Setting this true folds a
  // single screen-horizontal flip into the ingest so the canonical "original" is
  // NON-mirrored, matching web/Android. Applied in DISPLAY space, AFTER the rotate
  // and origin-snap (see uprightTransform step 3), so it is a true screen-
  // horizontal mirror (negate U) regardless of frame.rotation. It must NOT be
  // folded into the rotate via scaledBy: a source-space horizontal flip composed
  // with a 90/270 rotation is a diagonal reflection that swaps the U/V axes and
  // transposes flip-x with flip-y downstream.
  // Defaults to true (front-camera selfie mirror is the shipping case); flip to
  // false if a future rear-camera/source path is already un-mirrored. This is the
  // single mirror knob, the analogue of ROTATION_DIRECTION for the U axis.
  static let INGEST_MIRROR_X = true

  // Escape hatch only. The ingest is a CoreImage render, not a Metal pass, so it
  // does NOT incur the module's per-Metal-pass V-flip; this should stay false.
  // Set true ONLY if device testing shows the CoreImage ingest itself lands the
  // frame vertically inverted (in which case the per-pass parity story upstream
  // is wrong and should be re-derived, not patched here).
  static let INGEST_V_FLIP = false

  /// Normalize a frame rotation to {0, 90, 180, 270}.
  static func normalize(_ degrees: Int) -> Int { ((degrees % 360) + 360) % 360 }

  /// Does this frame rotation swap buffer dims into display dims?
  static func swaps(_ frameRotation: Int) -> Bool {
    let r = normalize(frameRotation)
    return r == 90 || r == 270
  }

  /// Display width for a buffer of `bufferWidth` x `bufferHeight` at `rotation`.
  static func displayWidth(bufferWidth: Int, bufferHeight: Int, rotation: Int) -> Int {
    swaps(rotation) ? bufferHeight : bufferWidth
  }

  /// Display height for a buffer of `bufferWidth` x `bufferHeight` at `rotation`.
  static func displayHeight(bufferWidth: Int, bufferHeight: Int, rotation: Int) -> Int {
    swaps(rotation) ? bufferWidth : bufferHeight
  }

  /// The CoreImage affine that maps the raw `sourceExtent` (buffer-space) image
  /// into a DISPLAY-UPRIGHT image whose origin is (0,0) and whose size is the
  /// display dims. The pipeline, in the order the pixels experience it:
  ///   1. rotate about the SOURCE center by ROTATION_DIRECTION * frameRotation,
  ///   2. snap the rotated bbox to the origin (now the upright DISPLAY rect),
  ///   3. THEN, in that display frame, fold in the de-mirror / V escape hatch.
  ///
  /// The mirror/V MUST be applied in display space (step 3), not folded into the
  /// rotate via `scaledBy` (step 1). `CGAffineTransform`'s builder methods PREPEND
  /// in point order, so an inline `t.scaledBy(...)` between the rotate and the
  /// inverse-center translate applies the scale to the point BEFORE the rotation,
  /// i.e. in SOURCE space. A horizontal flip in source space, composed with a
  /// 90/270 rotation, is a reflection across a DIAGONAL: it SWAPS the U and V axes
  /// of the produced texture. Downstream the transform shader negates U for flip-x
  /// and V for flip-y about an axis-aligned UV frame, so a U<->V-swapped original
  /// makes flip-x read as a vertical flip and flip-y as a horizontal mirror (the
  /// two ops transpose), and the debug grid lands rotated 90. Mirroring AFTER the
  /// rotation, about the DISPLAY center, is a clean screen-horizontal flip (negate
  /// U only) that leaves the axes aligned. (This is what the prior comment CLAIMED
  /// the code did; the composition order silently defeated it.)
  ///
  /// CoreImage is a bottom-left, CCW-positive coordinate system; this returns a
  /// transform to feed `CIImage.transformed(by:)` before `render(_:to:bounds:)`.
  static func uprightTransform(sourceExtent: CGRect, frameRotation: Int) -> CGAffineTransform {
    let r = normalize(frameRotation)
    let radians = ROTATION_DIRECTION * CGFloat(r) * .pi / 180.0

    // Step 1: rotate about the source center.
    let cx = sourceExtent.midX
    let cy = sourceExtent.midY
    var t = CGAffineTransform(translationX: cx, y: cy)
    t = t.rotated(by: radians)
    t = t.translatedBy(x: -cx, y: -cy)

    // Step 2: snap the rotated bbox to the origin so the content sits in the
    // positive quadrant; its size is now the upright DISPLAY rect.
    let rotatedBounds = sourceExtent.applying(t)
    t = t.concatenating(
      CGAffineTransform(translationX: -rotatedBounds.origin.x, y: -rotatedBounds.origin.y)
    )

    // Step 3: de-mirror / V escape hatch, IN DISPLAY SPACE. The upright content
    // now spans (0,0)..(displayW, displayH); reflect about the display center so
    // the mirror is a true screen-horizontal flip (negate U) regardless of
    // frame.rotation. Concatenated AFTER `t` (point order: rotate+snap first,
    // then this reflection), so it acts on the already-upright display frame.
    if INGEST_MIRROR_X || INGEST_V_FLIP {
      let displayW = rotatedBounds.width
      let displayH = rotatedBounds.height
      let dcx = displayW / 2
      let dcy = displayH / 2
      var mirror = CGAffineTransform(translationX: dcx, y: dcy)
      mirror = mirror.scaledBy(x: INGEST_MIRROR_X ? -1 : 1, y: INGEST_V_FLIP ? -1 : 1)
      mirror = mirror.translatedBy(x: -dcx, y: -dcy)
      t = t.concatenating(mirror)
    }

    return t
  }
}

File: android/src/main/java/com/simiancraft/kaleidoscope/gpu/Ingest.kt
// Ingest: the ONE place camera orientation is normalized.
//
// ===== Why this exists (read before "fixing" any effect's orientation) =====
//
// The camera hands us an OES external texture plus a 3x3 `transformMatrix`. That
// matrix bakes in sensor-correction, the selfie mirror, and crop, but it does
// NOT apply the display rotation: on a portrait phone the OES->2D passthrough,
// fed only `transformMatrix`, still lands a RAW LANDSCAPE frame in the FBO
// (device-confirmed 2026-05-24). Every effect then had to re-correct that
// rotation itself (transform ops conjugated in buffer space, the background
// composited pre-rotated). That per-effect cascade is what this module removes.
//
// `composedTexMatrix` folds the display rotation INTO the texture matrix, so the
// single OES->2D pass that BlurFactory, BackgroundImageFactory, and
// TransformFactory each run now lands a DISPLAY-UPRIGHT frame in the FBO. The
// FBO is sized with `displayWidth` / `displayHeight` (buffer dims swapped on a
// 90/270 frame), so the upright image fills it without clamping.
//
// After ingest the "original 2D" FBO is canonical: every downstream pass
// (transform mat2, background cover-fit, blur, the mask downsample/MediaPipe) reads
// an already-upright square [0,1] UV space and applies NO further orientation
// correction. Each factory therefore returns the frame with rotation 0; the
// pixels are already where the encoder/renderer expects an upright frame.
//
// ===== The composition =====
//
// The OES frag samples `uv = (uTexMatrix * vec4(vUv, 0, 1)).xy`, where `vUv`
// runs [0,1] across the destination (now display-upright) quad. We compose
//   M' = transformMatrix * Rot(theta about 0.5)
// so a destination display-UV is first rotated back into the pre-rotation space
// `transformMatrix` was authored against, then mapped to buffer UV. theta is
// `frame.rotation` scaled by ROTATION_DIRECTION.
//
// ===== Calibration (device verifies; may need one literal edit) =====
//
// ROTATION_DIRECTION below is the single sign that decides whether the display
// rotation is applied as +R or -R. If the person/background comes out rotated
// the wrong way (e.g. 90 off, or upside down on a 180 device), flip this sign.
// Nothing else in the orientation story should ever need touching.

package com.simiancraft.kaleidoscope.gpu

internal object Ingest {
  // The ONE orientation calibration knob. +1f applies the display rotation as
  // -frame.rotation about center (the geometric inverse, which is what maps a
  // display-upright destination UV back into the camera's pre-rotation UV
  // space); -1f applies +frame.rotation. Flip this sign if a portrait-device
  // screenshot shows the whole frame rotated the wrong way. This is the only
  // place orientation direction is decided.
  private const val ROTATION_DIRECTION = 1f

  /** Display width for a buffer of [bufferWidth] x [bufferHeight] at [frameRotation]. */
  fun displayWidth(bufferWidth: Int, bufferHeight: Int, frameRotation: Int): Int =
    if (swaps(frameRotation)) bufferHeight else bufferWidth

  /** Display height for a buffer of [bufferWidth] x [bufferHeight] at [frameRotation]. */
  fun displayHeight(bufferWidth: Int, bufferHeight: Int, frameRotation: Int): Int =
    if (swaps(frameRotation)) bufferWidth else bufferHeight

  /** Does this frame rotation swap buffer dims into display dims? */
  private fun swaps(frameRotation: Int): Boolean {
    val r = normalize(frameRotation)
    return r == 90 || r == 270
  }

  /**
   * Column-major 4x4 for the OES->2D pass's `uTexMatrix`, composing the camera
   * [transformMatrix] with the display rotation derived from [frameRotation].
   * Feed the result straight to glUniformMatrix4fv (already flat, length 16).
   *
   * M' = transformMatrix4 * Rot(theta about 0.5), where
   * theta = ROTATION_DIRECTION * frameRotation degrees.
   */
  fun composedTexMatrix(
    transformMatrix: android.graphics.Matrix,
    frameRotation: Int,
  ): FloatArray {
    val base = Egl.matrixToGl(transformMatrix)
    val rot = rotationAboutCenter(ROTATION_DIRECTION * normalize(frameRotation).toFloat())
    return multiplyColumnMajor4(base, rot)
  }

  /**
   * Column-major 4x4 that rotates the xy UV plane by [degrees] about (0.5, 0.5),
   * leaving z and w untouched. Translate-to-center, rotate, translate-back.
   */
  private fun rotationAboutCenter(degrees: Float): FloatArray {
    val rad = Math.toRadians(degrees.toDouble())
    val c = Math.cos(rad).toFloat()
    val s = Math.sin(rad).toFloat()
    // 2D rotation about 0.5 in (x, y):
    //   x' = c*(x-0.5) - s*(y-0.5) + 0.5
    //   y' = s*(x-0.5) + c*(y-0.5) + 0.5
    // As an affine 4x4 (column-major): linear block in the upper-left 2x2,
    // translation in the last column.
    val tx = 0.5f - 0.5f * c + 0.5f * s
    val ty = 0.5f - 0.5f * s - 0.5f * c
    return floatArrayOf(
      c, s, 0f, 0f,
      -s, c, 0f, 0f,
      0f, 0f, 1f, 0f,
      tx, ty, 0f, 1f,
    )
  }

  /** a * b, both column-major 4x4 flat arrays (length 16). Returns a*b. */
  private fun multiplyColumnMajor4(a: FloatArray, b: FloatArray): FloatArray {
    val out = FloatArray(16)
    for (col in 0 until 4) {
      for (row in 0 until 4) {
        var sum = 0f
        for (k in 0 until 4) {
          // column-major index: element(row, k) = a[k*4 + row]
          sum += a[k * 4 + row] * b[col * 4 + k]
        }
        out[col * 4 + row] = sum
      }
    }
    return out
  }

  private fun normalize(deg: Int): Int = ((deg % 360) + 360) % 360
}
